articles/cross_compilation_build_tools_design.md

```diff
@@ -0,0 +1,248 @@
+---
+layout: default
+title: Cross-compiling ROS 2 packages
+abstract:
+  This article is a design proposal for extending the build tools for ROS 2 to
+  simplify cross-compiling ROS 2 packages for platforms currently not supported
+  by ROS 2, or building on a host with different platforms than the target
+  platform. That later use case is particularly useful for speeding up build
+  times by building the code on a powerful development workstation, and then
+  deploying the binaries to a constrained resources hardware that uses the
+  target platform, thus making the development cycle much faster.
+author: >
+  [Thomas Moulard](https://github.com/thomas-moulard),
+  [Juan Rodriguez Hortala](https://github.com/juanrh)
+published: false
+---
+
+{:toc}
+
+# {{ page.title }}
+
+<div class="abstract" markdown="1">
+{{ page.abstract }}
+</div>
+
+Original Author: {{ page.author }}
+
+## Background
+
+The [cross-compilation tutorial](https://index.ros.org/doc/ros2/Tutorials/Cross-compilation)
+contains instructions for cross-compiling ROS 2 to unsupported platforms, using
+some configuration files for CMake and Docker that is setting maintained in the
+[ros2/cross_compile repository](https://github.com/ros2/cross_compile/).
+However, that tutorial doesn't provide instructions to build other ROS 2
+packages, but just for building a whole ROS 2 distribution. Ideally, we should
+be able to cross-compile a ROS package for ROS 2 by launching a single
+`colcon build` command. Finally, cross-compiling for other platforms besides
+ARM64 is not supported in the [ros2/cross_compile repository](https://github.com/ros2/cross_compile/).
+
+## Proposed Approach
+
+We propose continuing with the general approach of the cross-compilation
+tutorial, based on building a sysroot for the target platform using qemu and
+Docker, and then using `cmake-toolchains` with a toolchain file pointing
+`CMAKE_SYSROOT` to the sysroot location to compile with C and C++
+cross-compilers.
+
+On top of that we propose using a **colcon plugin `cc-build`** to wrap the
```

Comment:
Instead of adding a new verb, wouldn't it more appropriate to add this logic to the CMake extension for colcon? `cc-build` seems to assume that all packages that are to be crosscompiled use CMake.

Reply:
The tricky part is the sysroot management which is based on Docker. I think it's too brutal to directly pull a dependency on Docker in all cases when one is using `colcon`. Being in a plugin is a convenient way to get users to "opt-in".

Reply:
In addition to CMake and Python, `colcon` supports Gradle, Bazel, Cargo and probably more, hence why I believe this functionality should be part of each build plugin, because it's hard to cover them all in a single plugin. I'm thinking something like `colcon build --cmake-sysroot ... --python-sysroot ...` would work. Perhaps we could enforce that any new build tool that gets integrated into `colcon` has support for cross compiling.

Reply:
> Perhaps we could enforce that any new build tool that gets integrated into `colcon` has support for cross compiling.

This goes against one of the [stated software design goals](https://colcon.readthedocs.io/en/released/developer/design.html#software-design) of `colcon`:

    > Each functionality added should follow the principle “you don’t pay for what you don’t use”.

So if a plugin proving support for X doesn't need sysroot support I shouldn't need to implement it.

Reply:
Related to this see below where I suggest having the platform support builds embedded into docker images instead of on the host. So that you don't need to have the cross compilation resources on your local machine, just a docker dependency.